src/tests/breeds-get.test.ts

```diff
@@ -0,0 +1,43 @@
+import fetch from 'node-fetch'
+import { handler } from '../lambdas/breeds-get'
+
+const mockedFetch: jest.Mock = fetch as any
+
+jest.mock('node-fetch')
+
+describe('breeds-get handler', () => {
+  const mockPayload = {
+    message: {
+      briard: [],
+      buhund: ['norwegian'],
+      bulldog: ['boston', 'english', 'french'],
+    },
+    status: 'success',
+  }
+
+  const mockResponse = {
+    message: ['briard', 'norwegian buhund', 'boston bulldog', 'english bulldog', 'french bulldog'],
+    status: 'success',
+  }
+  beforeEach(() => {
+    mockedFetch.mockReturnValueOnce({
+      json: () => {
+        return mockPayload
+      },
+    })
+  })
+
+  it('returns data in correct format from fetch request', async () => {
+    const response = await handler()
+    expect(response).toMatchObject({ body: mockResponse })
+  })
+})
+
+describe('Reject Test', () => {
+  const mockResponse = { message: 'Something went wrong', statusCode: 500 }
+  it('returns error when request times out', async () => {
+    mockedFetch.mockRejectedValue({ status: 504, error: new Error('Service timeout') })
+    const response = await handler()
+    expect(response).toMatchObject(mockResponse)
+  })
```

Comment:
What we are looking to test here is the case when there is a timeout from the fetch call.  Really any error inside the handler would throw and cause this test to pass.  We want to isolate the timeout functionality in its own test case. 

It might be useful to review the [node fetch docs](https://github.com/node-fetch/node-fetch) for inspiration.

Reply:
Hey I just updated this. Can you let me know if this is more close to what you are looking for?

Reply:
Getting really close here.  A couple of things:

1. Can we move Abort controller into the handler please? 
2. We still have the same issue, where any error in the handler would cause this test to pass.  So might be better to think of this in terms of having two outcomes we test for: a) the handler times out b) the reduce function on [line 26](https://github.com/blazestudios23/lambda-bakcend-challenge/pull/1/files#diff-fbcc995aa5f538eb76f2e12a7f4cc4b9cae17bc12cc1d6dd28d138e79a4e266bR26) errors out for whatever reason. 

The goal is to isolate the timeout functionality in a single test.  We are almost there, just have to nudge it over the line.  Let me know if I can clarify anything further and please feel free to reach out with any additional questions.

Reply:
After about an hour of messing around I've determined that unfortunately the documentation for node-fetch is out dated and the example they give straight up doesn't work it throws typescript errors when trying to use the AbortController. See my edited code. 

I had to update the version of @types/node to get it to recognize the "globalThis" in the package.json file to version 14. 

I'm using node 14.19.3 through NVM. 

I get the following error:

```
TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
    src/lambdas/breeds-get.ts:27:65 - error TS2739: Type 'AbortSignal' is missing the following properties from type 'AbortSignal': addEventListener, removeEventListener, dispatchEvent, onabort

    27     const res = await fetch(`${DOG_BREEDS_ENDPOINT}list/all`, { signal: controller.signal })
```

Based off of the implementation in this documentation:
https://github.com/node-fetch/node-fetch#request-cancellation-with-abortsignal

Reply:
Definitely a mismatch in the documentation there.  That must have been changed fairly recently.  Just pulled your branch and it seems that if we import `AbortController` directly from its own package that resolves the typing issue:

```js
import AbortController from 'abort-controller'
```
That should also make mocking the response slightly more straightforward as well.  Hoping these docs are still relevant: https://github.com/node-fetch/node-fetch/blob/main/docs/ERROR-HANDLING.md 

Let me know if that doesn't roll us forward and perhaps we can schedule some time to pair on this. Thanks

Reply:
This is also great stuff finding this.  I am going to go through the challenge and make sure what we have lines up with the documentation rolling forward.  Appreciate you shouting this out.

Reply:
I was using the actual abort-controller before. 

I was logging out the error message so I could throw a different error if there's a timeout, but all I get is this:

```
onsole.log src/lambdas/breeds-get.ts:43
      TypeError: Cannot read properties of undefined (reading 'json')
          at handler (/Users/andrewobrigewitsch/Downloads/lambda-backend-challenge-master/src/lambdas/breeds-get.ts:27:39)
          at processTicksAndRejections (node:internal/process/task_queues:96:5)
          at Object.<anonymous> (/Users/andrewobrigewitsch/Downloads/lambda-backend-challenge-master/src/tests/breeds-get.test.ts:40:22)
    console.log src/lambdas/breeds-get.ts:43
```

Do I also need to manually set a mock reject like this?

```
mockedFetch.mockRejectedValue({ status: 504, error: new Error('service timed out') })
```

Reply:
Rather than mocking the fetch request to test the timeout, we can mock the AbortController package.  That way we can specifically test that when signal aborts via that package a specific string is rendered in the error message and that will allow us to isolate the timeout functionality on its own.  

Hopefully that makes sense. Feel like we are really close here.  Let me know if I can clarify further.  Cheers

Reply:
That does make sense, but it looks like that package has change too from all the examples I'm seeing on the web and I'm not actually able to get any output from the signal besides undefined.

Reply:
I'm going to push all the code I've added to test it just so you can see what I'm doing.

Reply:
I also had to upgrade TS, JEST-TS and JEST packages because they where throwing warnings.

Reply:
It's probably best to just jump on a call tomorrow to go over it.

Reply:
Ok sounds good, I am going to pull your latest branch after standup and poke around a bit. We can definitely hop on a call to sort this out. Will follow up in a couple hours. Cheers

Reply:
Ok, so I think we can work within the setup that we have here.  If you check the [node-fetch docs for error handling](https://github.com/node-fetch/node-fetch/blob/main/docs/ERROR-HANDLING.md).  AbortError throws with this property:

```js
if (error.name === 'AbortError') {
  console.log('request was aborted');
}
```

Which would allow us to construct a test like so as the fetch call would throw with that error:

```js
const error = new Error('Request timeout')
error.name = 'AbortError'
mockedFetch.mockRejectedValue(error)
```

Just need to set up the catch block in the handler to conditionally return an appropriate response.  Please do keep the `error: unknown` as that adds type safety to what we are doing.  We can use the error instance and `name` properties to lock down the flow. 

I do have free time today as well, if you want to jump on call and discuss in detail. Just let me know when you are free to chat.  Cheers